### PR TITLE
feat: add profile switch in cli

### PR DESCRIPTION
**Pull request by ahabhgk**
## Summary

- add js side cpuprofile in `@rspack/cli`
- expose `experimental_registerGlobalTrace` and `experimental_cleanupGlobalTrace` from `@rspack/core`, so `@rspack/cli`, modern.js, or other `@rspack/core` users can use it to enable rust side tracing
- add `RSPACK_PROFILE` in `@rspack/cli`:

```
`RSPACK_PROFILE='TRACE=filter=trace&output=./rspack.trace&layer=chrome|JSCPU=output=./rspack.jscpuprofile' npx rspack`
                       ^----------------------------------------------: querystring syntax trace options
                                                                      ^: | is a delimiter for different profile options
                                                                       ^---------------------------------: querystring syntax js cpuprofile options
                       ^-----------: trace filter, default to `trace`, more syntax: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#example-syntax
                                    ^--------------------: trace output, `stderr`, `stdout`, or a file path, default to `./rspack.trace` for layer `chrome` and default to `stdout` for layer `logger`
                                                          ^-----------: trace layer, `chrome` or `logger`, default to `chrome`
                                                                             ^---------------------------: js cpuprofile output, `stderr`, `stdout`, or a file path, default to `./rspack.jscpuprofile`

`RSPACK_PROFILE='TRACE=filter=trace&output=./rspack.trace&layer=chrome' npx rspack`: only enable trace and disable js cpuprofile

`RSPACK_PROFILE=TRACE npx rspack`: only enable trace and disable js cpuprofile, and use default options of trace

`RSPACK_PROFILE='JSCPU=output=./rspack.jscpuprofile' npx rspack`: only enable js cpuprofile and disable trace

`RSPACK_PROFILE=JSCPU npx rspack`: only enable js cpuprofile and disable trace, and use default options of js cpuprofile

`RSPACK_PROFILE=ALL npx rspack`: enable both trace and js cpuprofile, and use default options of trace and js cpuprofile

`RSPACK_PROFILE=[rspack_node,rspack_core] npx rspack`: enable both trace and js cpuprofile, but customize trace filter
```

## Test Plan

- `packages/rspack-cli/tests/build/profile`

### CODE REVIEW

**Comment by hardfist** [comment]
`RSPACK_PROFILE=TRACE=trave` this seems weird and I can't get its meaning?

**Comment by ahabhgk** [comment]
Sorry, `trace`, [a tracing level as a tracing filter](https://github.com/web-infra-dev/rspack/blob/0ff053b9d378dfed23e7a76ce71b4227aa11f7f6/crates/rspack_tracing/src/lib.rs#L49)

**Comment by hardfist** [comment]
> Sorry, `trace`, [a tracing level as a tracing filter](https://github.com/web-infra-dev/rspack/blob/0ff053b9d378dfed23e7a76ce71b4227aa11f7f6/crates/rspack_tracing/src/lib.rs#L49)

it's not about typo, it's about the format of cli args,"a=b=c" seems not a good and readable args format

**Comment by ahabhgk** [comment]
`=` is learned from `querystring`, we can use other style if you have better idea

**Comment by hardfist** [comment]
> `=` is learned from `querystring`, we can use other style if you have better idea

I suggest we follow rustc style https://rustc-dev-guide.rust-lang.org/tracing.html#query-level-filters
`RSPACK_PROFILE=[code_splitting,bundle_splitting]` if we use `[]` here, it just means we use trace profile and no need for extra `TRACE=` to annotate we use trace profile
and we use the same the filter syntax as rustc, so it's easy to learn

File: packages/rspack-cli/src/utils/profile.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+};
+export type RustTrace = {
+	kind: "rust trace";
+	filter: string;
+	output: string;
+};
+
+export type ProfileOptions = JavascriptCPUProfile | RustTrace;
+
+const defaultRustTraceOutput = `./rspack.trace`;
```
Comment (hardfist) [inline packages/rspack-cli/src/utils/profile.ts:17]: should we put it into cwd or in the output_dir? nextjs seems put it into output_dir

File: crates/node_binding/src/lib.rs [hunk 2]
```diff
@@ -308,3 +307,44 @@ impl Rspack {
...
+  filter: String,
+  layer: String,
+  output: String,
+) -> Result<()> {
+  CUSTOM_TRACE_SUBSCRIBER.get_or_init(|| {
+    let guard = match layer.as_str() {
+      "chrome" => {
+        rspack_tracing::enable_tracing_by_env_with_chrome_layer(&filter, Path::new(&output))
+      }
+      "logger" => {
```
[the remark this answers is not in the document]
Comment (ahabhgk) [reply crates/node_binding/src/lib.rs:332]: I'm not sure, @hyf0 may use it

File: packages/rspack-cli/src/utils/profile.ts (hunk 1 above)
Comment (ahabhgk) [reply packages/rspack-cli/src/utils/profile.ts:17]: current is cwd

File: crates/node_binding/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (hardfist) [reply crates/node_binding/src/lib.rs:332]: then let's keep it now, @hyf0 can you explain why you still need the logger layer

[the remark this answers is not in the document]
Comment (hyf0) [reply crates/node_binding/src/lib.rs:332]: You could remove it if necessary.

File: packages/rspack-cli/src/rspack-cli.ts [hunk 3]
```diff
@@ -147,6 +152,21 @@ export class RspackCLI {
 					}
 				});
 			}
+			if (process.env.RSPACK_PROFILE) {
+				resolveProfile(process.env.RSPACK_PROFILE).forEach(profileOptions => {
+					if (profileOptions.kind === "rust trace") {
```
Comment (9aoy) [inline packages/rspack-cli/src/rspack-cli.ts:157]: It seems that spaces are rarely seen, usually connected with underscores or dashes?

File: packages/rspack/src/index.ts [hunk 4]
```diff
@@ -11,6 +11,10 @@ export { cachedCleverMerge as cleverMerge } from "./util/cleverMerge";
 export { BannerPlugin } from "./lib/BannerPlugin";
 export { EnvironmentPlugin } from "./lib/EnvironmentPlugin";
 export { LoaderOptionsPlugin } from "./lib/LoaderOptionsPlugin";
+export {
+	registerGlobalTrace as experimental_registerGlobalTrace,
```
[the remark this answers is not in the document]
Comment (ahabhgk) [reply packages/rspack/src/index.ts:15]: Webpack don't have this functionality, we can use [`options.profile = true`](https://webpack.js.org/configuration/other-options/#profile) to do the extra work (enable RSPACK_PROFILE functionality)

Or we ship it in experimental prefix as experimental api and change it anytime

[the remark this answers is not in the document]
Comment (hardfist) [reply packages/rspack/src/index.ts:15]: experimental sounds good to me

**Review by hardfist — APPROVED** [review]

**Pull request merged**